authentication.go

```diff
@@ -178,3 +204,75 @@ func (s *AuthenticationService) GetCurrentUser() (*Session, error) {
+// in a JiraAccessToken struct so that it can be cached by the client application if desired (oauth2 tokens expire
+// after 15 min by default); else, this can be called multiple times to retrieve a new access token.
+//
+// oauthClientId, userKey, and sharedSecret are returned by Jira during the installation of the Add-on
+// scope is the scope to place on the AccessToken returned (e.g., "READ", "READ WRITE", "READ ACT_AS_USER")
+// See: https://developer.atlassian.com/cloud/jira/software/oauth-2-jwt-bearer-token-authorization-grant-type
+func (s *AuthenticationService) GetOauth2AccessToken(oauthClientId, userKey, scope string, sharedSecret []byte) (*JiraAccessToken, error) {
+	// create JSON Web Token
+	jiraAuthURL := "https://auth.atlassian.io"
+	jwtStr, err := createJWT(oauthClientId, s.client.baseURL.String(), jiraAuthURL, userKey, sharedSecret)
+	if err != nil {
+		return nil, fmt.Errorf("Error creating JSON Web Token : %s", err)
+	}
+
+	// Now make a request to the Jira auth server for an access token
+	oauth2Route := fmt.Sprintf("%s/oauth2/token", jiraAuthURL)
+	grantType := "urn:ietf:params:oauth:grant-type:jwt-bearer"
+	form := url.Values{"grant_type": {grantType}, "assertion": {jwtStr}, "scope": {scope}}
+	formBody := strings.NewReader(form.Encode())
+	req, err := http.NewRequest("POST", oauth2Route, formBody)
+	if err != nil {
+		return nil, fmt.Errorf("Error creating Jira Auth Server request : %s", err)
+	}
+	req.Header.Set("Content-Length", strconv.FormatInt(int64(len(form.Encode())), 10))
+	req.Header.Set("Content-Type", "application/x-www-form-urlencoded; charset=utf-8")
+	req.Header.Set("alg", "HS256")
+	req.Header.Set("typ", "JWT")
+	client := &http.Client{}
+	resp, err := client.Do(req)
+	if err != nil {
+		return nil, fmt.Errorf("Error getting Oauth2 access token from %s : %s", oauth2Route, err)
+	}
+
+	// Finally, decode the JSON from the response
+	defer resp.Body.Close()
+	token := new(JiraAccessToken)
+	if err := json.NewDecoder(resp.Body).Decode(token); err != nil {
+		return nil, fmt.Errorf("Error decoding JSON from response %+v : %s", resp, err)
+	}
+
+	// set oauth type before leaving
+	s.SetOauth(token.AccessToken)
+
+	return token, nil
+}
+
+// Creates a JSON Web Token that will be used as an assertion to get an Oauth2 access token
+// See: https://developer.atlassian.com/cloud/jira/software/understanding-jwt
+func createJWT(oauthClientId, instanceURL, jiraAuthURL, userKey string, sharedSecret []byte) (string, error) {
+	jwt := jwt.NewWithClaims(jwt.SigningMethodHS256, jwt.MapClaims{
+		"iss": fmt.Sprintf("urn:atlassian:connect:clientid:%s", oauthClientId),
+		"tnt": instanceURL,
+		"sub": fmt.Sprintf("urn:atlassian:connect:userkey:%s", userKey),
+		"aud": jiraAuthURL,
+		"iat": time.Now().Unix(),
+		"exp": time.Now().Unix() + 10, // expires in 10 seconds
+	})
+
+	// sign the jwt using the signing method and the shared secret
+	if signingStr, err := jwt.SigningString(); err != nil {
+		return "", err
+	} else {
```

Comment:
Not sure this needs an `else` since you've called `return` early on the previous error.

Reply:
`signingStr` loses scope outside of the if / else block and I only need it for the subsequent call.